internal/mtls/ca.go

```diff
@@ -0,0 +1,187 @@
+package mtls
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"net/http"
+
+	"github.com/hashicorp/go-multierror"
+	"github.com/jakub-dzon/k4e-operator/internal/yggdrasil"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilrand "k8s.io/apimachinery/pkg/util/rand"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	regClientSecretNameRandomLen = 10
+	regClientSecretNamePrefix    = "reg-client-ca"
+	regClientSecretLabelKey      = "reg-client-ca"
+)
+
+// CAProvider The main reason to have an interface here is to be able to extend this to
+// future Cert providers, like:
+// - Vault
+// - Acme protocol
+// Keeping as an interface, so in future users can decice.
+type CAProvider interface {
+	GetName() string
+	GetCACertificate() (*CertificateGroup, error)
+	CreateRegistrationCertificate(name string) (map[string][]byte, error)
+}
+
+type TLSConfig struct {
+	config           *tls.Config
+	client           client.Client
+	caProvider       []CAProvider
+	Domains          []string
+	LocalhostEnabled bool
+	namespace        string
+}
+
+func NewMTLSConfig(client client.Client, namespace string, domains []string, localhostEnabled bool) *TLSConfig {
+	config := &TLSConfig{
+		config:           nil,
+		client:           client,
+		Domains:          domains,
+		namespace:        namespace,
+		LocalhostEnabled: localhostEnabled,
+	}
+
+	// Secret providers here
+	secretProvider := NewCASecretProvider(client, namespace)
+	config.caProvider = append(config.caProvider, secretProvider)
+
+	return config
+}
+
+// @TODO mainly used for testing, maybe not needed at all
+func (conf *TLSConfig) SetCAProvider(caProviders []CAProvider) {
+	conf.caProvider = caProviders
+}
+
+func (conf *TLSConfig) InitCertificates() (*tls.Config, []*x509.Certificate, error) {
+	if len(conf.caProvider) == 0 {
+		return nil, nil, fmt.Errorf("no CA provider is set")
+	}
+
+	var errors error
+	caCerts := []*CertificateGroup{}
+
+	CACertChain := []*x509.Certificate{}
+	caCertPool := x509.NewCertPool()
+
+	for _, caProvider := range conf.caProvider {
+		caCert, err := caProvider.GetCACertificate()
+		if err != nil {
+			errors = multierror.Append(errors, fmt.Errorf(
+				"cannot get CA certificate for provider %s: %v",
+				caProvider.GetName(), err))
+			continue
+		}
+
+		caCerts = append(caCerts, caCert)
+		CACertChain = append(CACertChain, caCert.GetCert())
+		caCertPool.AppendCertsFromPEM(caCert.certPEM.Bytes())
+	}
+
+	if errors != nil {
+		return nil, nil, errors
+	}
+
+	if len(caCerts) == 0 {
+		return nil, nil, fmt.Errorf("Cannot get any CA certificate")
```

Comment:
as a general comment for all `fmt.Errorf` calls, according to the guidelines:
"Error strings should not be capitalized (unless beginning with proper nouns or acronyms) or end with punctuation"

https://github.com/golang/go/wiki/CodeReviewComments#error-strings